File: library/src/main/java/com/readystatesoftware/chuck/api/RetentionManager.java
package com.readystatesoftware.chuck.api;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import com.readystatesoftware.chuck.internal.data.repository.ChuckerRepositoryProvider;

import java.util.concurrent.TimeUnit;

public class RetentionManager {

    private static final String LOG_TAG = "Chucker";
    private static final String PREFS_NAME = "chucker_preferences";
    private static final String KEY_LAST_CLEANUP = "last_cleanup";

    private static long lastCleanup;

    private final long period;
    private final long cleanupFrequency;
    private final SharedPreferences prefs;

    public RetentionManager(Context context, ChuckCollector.Period retentionPeriod) {
        period = toMillis(retentionPeriod);
        prefs = context.getSharedPreferences(PREFS_NAME, 0);
        cleanupFrequency = (retentionPeriod == ChuckCollector.Period.ONE_HOUR) ?
                TimeUnit.MINUTES.toMillis(30) : TimeUnit.HOURS.toMillis(2);
    }

    public synchronized void doMaintenance() {
        if (period > 0) {
            long now = System.currentTimeMillis();
            if (isCleanupDue(now)) {
                Log.i(LOG_TAG, "Performing data retention maintenance...");
                deleteSince(getThreshold(now));
                updateLastCleanup(now);
            }
        }
    }

    private long getLastCleanup(long fallback) {
        if (lastCleanup == 0) {
            lastCleanup = prefs.getLong(KEY_LAST_CLEANUP, fallback);
        }
        return lastCleanup;
    }

    private void updateLastCleanup(long time) {
        lastCleanup = time;
        prefs.edit().putLong(KEY_LAST_CLEANUP, time).apply();
    }

    private void deleteSince(long threshold) {
        ChuckerRepositoryProvider.it().deleteOldTransactions(threshold);
        ChuckerRepositoryProvider.it().deleteOldThrowables(threshold);
    }

    private boolean isCleanupDue(long now) {
        return (now - getLastCleanup(now)) > cleanupFrequency;
    }

    private long getThreshold(long now) {
        return (period == 0) ? now : now - period;
    }

    private long toMillis(ChuckCollector.Period period) {
        switch (period) {
            case ONE_HOUR:
                return TimeUnit.HOURS.toMillis(1);
            case ONE_DAY:
                return TimeUnit.DAYS.toMillis(1);
            case ONE_WEEK:
                return TimeUnit.DAYS.toMillis(7);
            case FOREVER:
            default:
                return 0;
        }
    }
}

File: library/src/main/kotlin/com/chuckerteam/chucker/api/RetentionManager.kt
package com.chuckerteam.chucker.api

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit
import com.chuckerteam.chucker.api.RetentionManager.Period
import com.chuckerteam.chucker.internal.data.repository.RepositoryProvider
import com.chuckerteam.chucker.internal.support.Logger
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.concurrent.TimeUnit

/**
 * Class responsible of holding the logic for the retention of your HTTP transactions.
 * You can customize how long data should be stored here.
 * @param context An Android Context
 * @param retentionPeriod A [Period] to specify the retention of data. Default 1 week.
 */
@Suppress("MagicNumber")
public class RetentionManager @JvmOverloads constructor(
    context: Context,
    retentionPeriod: Period = Period.ONE_WEEK
) {

    // The actual retention period in milliseconds (default to ONE_WEEK)
    private val period: Long = toMillis(retentionPeriod)

    // How often the cleanup should happen
    private val cleanupFrequency: Long
    private val prefs: SharedPreferences by lazy {
        context.getSharedPreferences(PREFS_NAME, 0)
    }
    private val maintenanceMutex = Mutex()

    init {
        cleanupFrequency = if (retentionPeriod == Period.ONE_HOUR) {
            TimeUnit.MINUTES.toMillis(30)
        } else {
            TimeUnit.HOURS.toMillis(2)
        }
    }

    /**
     * Call this function to check and eventually trigger a cleanup.
     * Please note that this method is not forcing a cleanup.
     */
    internal suspend fun doMaintenance() = maintenanceMutex.withLock {
        if (period > 0) {
            val now = System.currentTimeMillis()
            if (isCleanupDue(now)) {
                Logger.info("Performing data retention maintenance...")
                deleteSince(getThreshold(now))
                updateLastCleanup(now)
            }
        }
    }

    private fun getLastCleanup(fallback: Long): Long {
        if (lastCleanup == 0L) {
            lastCleanup = prefs.getLong(KEY_LAST_CLEANUP, fallback)
        }
        return lastCleanup
    }

    private fun updateLastCleanup(time: Long) {
        lastCleanup = time
        prefs.edit { putLong(KEY_LAST_CLEANUP, time) }
    }

    private suspend fun deleteSince(threshold: Long) {
        RepositoryProvider.transaction().deleteOldTransactions(threshold)
    }

    private fun isCleanupDue(now: Long) = now - getLastCleanup(now) > cleanupFrequency

    private fun getThreshold(now: Long) = if (period == 0L) now else now - period

    private fun toMillis(period: Period): Long {
        return when (period) {
            Period.ONE_HOUR -> TimeUnit.HOURS.toMillis(1)
            Period.ONE_DAY -> TimeUnit.DAYS.toMillis(1)
            Period.ONE_WEEK -> TimeUnit.DAYS.toMillis(7)
            Period.FOREVER -> 0
        }
    }

    public enum class Period {
        /** Retain data for the last hour. */
        ONE_HOUR,

        /** Retain data for the last day. */
        ONE_DAY,

        /** Retain data for the last week. */
        ONE_WEEK,

        /** Retain data forever. */
        FOREVER
    }

    private companion object {
        private const val PREFS_NAME = "chucker_preferences"
        private const val KEY_LAST_CLEANUP = "last_cleanup"
        private var lastCleanup: Long = 0
    }
}
